### PR TITLE
GATEWAY-2129 and GATEWAY-3028

### PR DESCRIPTION
**Pull request by bhumphrey**
Refactor of outbound document query to use the orchestration approach.

### CODE REVIEW

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/common/connectionmanager/dao/AssigningAuthorityHomeCommunityMappingDAO.java [hunk 1]
```diff
@@ -94,29 +70,22 @@ public String getAssigningAuthority(String homeCommunityId) {
      * @param homeCommId
      * @return List
      */
+    @SuppressWarnings("unchecked")
     public List<String> getAssigningAuthoritiesByHomeCommunity(String homeCommunityId) {
-        log.debug("-- Begin AssigningAuthorityHomeCommunityMappingDAO.getAssigningAuthoritiesByHomeCommunity() ---");
+        log.trace("-- Begin AssigningAuthorityHomeCommunityMappingDAO.getAssigningAuthoritiesByHomeCommunity() ---");
```
Comment (tparnell) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/common/connectionmanager/dao/AssigningAuthorityHomeCommunityMappingDAO.java:75]: Mismatched log levels for entry and exit.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/connectmgr/ConnectionManager.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+import gov.hhs.fha.nhinc.nhinclib.NhincConstants.UDDI_SPEC_VERSION;
+
+import java.util.List;
+import java.util.Set;
+
+import org.uddi.api_v3.BusinessEntity;
+
+public interface ConnectionManager {
+
+    /**
```
Comment (tparnell) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/connectmgr/ConnectionManager.java]: remove

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+                    NhincConstants.DOC_QUERY_SERVICE_NAME)) {
+                patientCorrelationReq.getTargetHomeCommunity().add(urlInfo.getHcid());
+            }
+
+            RetrievePatientCorrelationsResponseType results = patientCorrelationProxyFactory
+                    .getPatientCorrelationProxy().retrievePatientCorrelations(
+                            pixRetrieveBuilder.createPixRetrieve(patientCorrelationReq), assertion);
+
+            List<QualifiedSubjectIdentifierType> subIdList = new ArrayList<QualifiedSubjectIdentifierType>();
+            List<II> identities = Collections.EMPTY_LIST;
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java]: Isn't this the non-type safe version?  Versus Collections.emptyList();

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    }
+
+    /**
+     * @param request
+     * @param sTargetHomeCommunityId
+     * @return
+     */
+    protected AdhocQueryRequest setTargetHomeCommunityId(AdhocQueryRequest request, String sTargetHomeCommunityId) {
+        if (NullChecker.isNotNullish(sTargetHomeCommunityId)) {
+            if (!(sTargetHomeCommunityId.startsWith("urn:oid:"))) {
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java]: I believe there's a constant somewhere for that string.  Please find + use.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        newQuery.setLid(currentQuery.getLid());
+        newQuery.setName(currentQuery.getName());
+        newQuery.setObjectType(currentQuery.getObjectType());
+        newQuery.setQueryExpression(currentQuery.getQueryExpression());
+        newQuery.setStatus(currentQuery.getStatus());
+        newQuery.setVersionInfo(currentQuery.getVersionInfo());
+        newRequest.setAdhocQuery(newQuery);
+
+        newRequest.setComment(request.getComment());
+        // newRequest.setFederated(request.isFederated());
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java]: Commented out??

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryAggregator.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+    /**
+     * @param individualResponse
+     * @param aggregatedResponse
+     */
+    protected void aggregateRegistryErrors(AdhocQueryResponse aggregatedResponse, AdhocQueryResponse individualResponse) {
+        // add any registry errors
+        RegistryErrorList registryErrorList = individualResponse.getRegistryErrorList();
+        if (registryErrorList != null
+                && registryErrorList.getRegistryError() != null
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryAggregator.java]: This can never be non-null.  JAXB populates upon call.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryAggregator.java [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+ * 
+ */
+public class OutboundDocQueryAggregator implements NhinAggregator {
+
+    /**
+     * @param individualResponse
+     * @param aggregatedResponse
+     */
+    protected void aggregateRegistryErrors(AdhocQueryResponse aggregatedResponse, AdhocQueryResponse individualResponse) {
+        // add any registry errors
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryAggregator.java]: These comments in the individual protected methods seem unnecessary.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryAggregator.java [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+    /**
+     * @param individualResponse
+     * @param aggregatedResponse
+     */
+    protected void aggregateSlotlistResponse(AdhocQueryResponse aggregatedResponse,
+            AdhocQueryResponse individualResponse) {
+        // add any slotlist response data
+        SlotListType individualResponseSlotList = individualResponse.getResponseSlotList();
+        if (individualResponseSlotList != null && individualResponseSlotList.getSlot() != null
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryAggregator.java]: getSlot will never be null.  JAXB auto-create.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryProcessor.java [hunk 9]
```diff
@@ -338,142 +313,211 @@ public OutboundDocQueryOrchestratable_a1 processError_a1(OutboundDocQueryOrchest
...
+            throw new Exception(
+                    "EntityDocQueryProcessor::aggregateResponse_a1 received a null AdhocQueryResponse response!!!");
+        }
+    }
+
+    /**
+     * @param individualResponse
+     * @param aggregatedResponse
+     */
+    protected void aggregateRegistryErrors(AdhocQueryResponse individualResponse, AdhocQueryResponse aggregatedResponse) {
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryProcessor.java:383]: This method (and a couple that follow) seem to be repeat code from earlier in the diff.  Can they be extracted into a common class?

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQueryHelper.java [hunk 10]
```diff
@@ -272,10 +132,10 @@ public String getUniquePatientId(List<SlotType1> slotList) {
 
         // For each slot process each of the Patient Id slots
         for (SlotType1 slot : slotList) {
-            if (slot.getName().equalsIgnoreCase(NhincConstants.DOC_QUERY_XDS_PATIENT_ID_SLOT_NAME)) {
+            if (NhincConstants.DOC_QUERY_XDS_PATIENT_ID_SLOT_NAME.equalsIgnoreCase(slot.getName())) {
                 if (slot.getValueList() != null && NullChecker.isNotNullish(slot.getValueList().getValue())
                         && NullChecker.isNotNullish(slot.getValueList().getValue().get(0))) {
-                    uniquePatientId = PatientIdFormatUtil.stripQuotesFromPatientId(slot.getValueList().getValue()
+                    uniquePatientId = PatientIdFormatUtil.parsePatientId(slot.getValueList().getValue()
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQueryHelper.java:138]: What is going on with this change?  It does not seem related to other changes.

File: Product/Production/Services/DocumentQueryCore/src/test/java/gov/hhs/fha/nhinc/docquery/entity/FanoutServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+
+    @Test
+    public void createChildRequestsSingle() throws ConnectionManagerException {
+        AdhocQueryRequest adhocQueryRequest = createRequest(createSlotList());
+        NhinTargetCommunitiesType targets = createNhinTargetCommunites();
+        AssertionType assertion = new AssertionType();
+
+        List<UrlInfo> urlInfoList = new ArrayList<UrlInfo>();
+        urlInfoList.add(createUrlInfo("4.4", ""));
+        // UrlInfo urlInfo1 = new UrlInfo();
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/test/java/gov/hhs/fha/nhinc/docquery/entity/FanoutServiceTest.java]: Commented out.

File: Product/Production/Services/DocumentQueryCore/src/test/java/gov/hhs/fha/nhinc/docquery/entity/FanoutServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    private void assertSame(AdhocQueryType query1, AdhocQueryType query2) {
+        List<SlotType1> slotList1 = query2.getSlot();
+
+        List<SlotType1> slotList2 = query1.getSlot();
+
+        assertSame(slotList1, slotList2);
+
+        // query2.getClassification().addAll(query1.getClassification());
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/test/java/gov/hhs/fha/nhinc/docquery/entity/FanoutServiceTest.java]: Commented out.

File: Product/Production/Services/DocumentQueryCore/src/test/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQueryTest.java [hunk 13]
```diff
@@ -6,277 +6,250 @@
...
+        targetCommunity.setRegion(region);
+        targetCommunity.setList(list);
+        return targetCommunity;
+    }
+    
+    private NhinTargetCommunitiesType createNhinTargetCommunites() {
+        return createTargetCommunities(createTargetCommunity("4.4", "US-FL", "Unimplemented"));
     }
+    
+//    @Test
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/test/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQueryTest.java]: Commented out.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryProcessor.java (hunk 9 above)
Comment (bhumphrey) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryProcessor.java:383]: this class should be removed.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQueryHelper.java (hunk 10 above)
Comment (bhumphrey) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQueryHelper.java:138]: This is correcting a mistake that was made earlier.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+            request.getAdhocQuery().setHome(sTargetHomeCommunityId);
+        }
+        return request;
+    }
+
+    /**
+     * @param request AdhocQuery request received.
+     * @return AdhocQUery Request.
+     */
+    protected AdhocQueryRequest cloneRequest(AdhocQueryRequest request) {
```
Comment (akong) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java:229]: I think a better way of doing deep copies of JAXBObject is to marshall to XML and then unmarshall it back.  

See:
http://stackoverflow.com/questions/930840/how-do-i-clone-a-jaxb-object
https://gist.github.com/821410 

Or in our code:
AsyncMessageProcessHelper::copyAssertionTypeObject()

With that said, Paul had problems with it and hence why he did it manually like this.  Might be worth a shot to try it so we don't have to update this method if the schema changed.

Comment (bhumphrey) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/AggregationService.java:229]: created [GATEWAY-3207](https://issues.connectopensource.org/browse/GATEWAY-3207)
